ExclusiveSyncPoint

belliottsmith opened the pull request:
Introduce a special kind of `ExclusiveSyncPoint` transaction that invalidates all `TxnId` lower than it that are not witnessed by it. This permits bootstrapping nodes to ignore all `TxnId` older than this when starting their log.

aweisberg commented on accord-core/src/main/java/accord/impl/InMemoryCommandStore.java:
```diff
@@ -374,9 +528,9 @@ public void run()
         }
 
-        class AsyncState extends State implements SafeCommandStore
+        class AsyncState extends InMemoryCommandStore.State implements SafeCommandStore
```
Is removing the import an improvement?

belliottsmith replied:
I think this is just Java 8's compilation woes, sometimes we have to rejig imports until it decides things are fine again. Just noise.

aweisberg replied:
Yikes!

aweisberg commented on accord-core/src/main/java/accord/local/CommandStore.java:
```diff
@@ -54,6 +63,36 @@ public int id()
+    Timestamp preaccept(TxnId txnId, Seekables<?, ?> keys, SafeCommandStore safeStore)
+    {
+        NodeTimeService time = safeStore.time();
+        boolean isExpired = agent().isExpired(txnId, safeStore.time().now());
```
Maybe rename `isExpired` to `shouldReject`?

belliottsmith replied:
We use it on both replica and coordinator, and on coordinator that perhaps isn't quite right. Though we could have different criteria for each, or we could stop checking it on coordinator and just check if any of the votes were to reject it,

aweisberg replied:
I am just reacting to `isExpired` being reused for the result of `rejectBefore`
If you want to call both of them expiration that's pretty reasonable.

belliottsmith replied:
Oh, I see. I thought you meant the method, but you mean the local variable. Makes perfect sense.

aweisberg commented on accord-core/src/main/java/accord/coordinate/CoordinateSyncPoint.java:
```diff
@@ -0,0 +1,93 @@
+    {
+        Deps deps = Deps.merge(successes, ok -> ok.deps);
+        Timestamp executeAt = foldl(successes, (ok, prev) -> mergeMax(ok.witnessedAt, prev), Timestamp.NONE);
+        if (executeAt.isRejected())
```
Should we expire?

belliottsmith replied:
Do you mean should we check the `agent.isExpired` thing? Since this has no user visible impact, and we presumably want it for some process reason it didn't seem valuable.

aweisberg replied:
Looks like we expire requests so if the transaction is stuck it will eventually generate an error so yeah seems fine to not align with a specific timeout just lack of progress.

I do wonder if it could get wedged in say the transaction logic, but then a retry wouldn't help either.

aweisberg commented on accord-core/src/main/java/accord/coordinate/CoordinateSyncPoint.java:
```diff
@@ -0,0 +1,93 @@
+            // that *may* execute before their txnId, so once these dependencies apply we can say that any action that
+            // awaits these dependencies applies after them. In the case of ExclusiveSyncPoint, we additionally guarantee
+            // that no lower TxnId can later apply.
+            new Propose<SyncPoint>(node, tracker.topologies(), Ballot.ZERO, txnId, txn, route, deps, txnId, this)
```
Anonymously subclassing `Propose` to override `onAccepted` is moderately weird, but sure. For `ProposeAndExecute` there is a class.

belliottsmith replied:
Happy to introduce an inner class. I did previously have a static method I could pass a callback to that would create an anonymous class to wrap the callback but that felt wrong.

aweisberg replied:
I don't really care. It's not hard to understand I just noted we have 3 ways to get at different types of proposals (Invalidate, ProposeAndExecute, and this).

belliottsmith replied:
I guess we could put `ProposeAndExecute` as an inner class. I don't think of this as three different kinds of Propose though, but as two different behaviours conducted after `Propose`, and one special kind of propose (invalidation), as this is behavioural different at the replica, rather than in action following. 

We could chain a layer of indirection with callbacks, but it seemed clearer to me this way. I don't have a super strong feeling, but would prefer to retain the symmetry of `Invalidate` with other classes

aweisberg commented on accord-core/src/main/java/accord/coordinate/CoordinateSyncPoint.java:
```diff
@@ -0,0 +1,93 @@
+                @Override
+                void onAccepted()
+                {
+                    node.send(tracker.nodes(), id -> new Apply(id, tracker.topologies(), tracker.topologies(), txnId.epoch(), txnId, route, txnId, deps, txn.execute(txnId, null), txn.result(txnId, null)));
```
We send the apply to avoid having this transaction block others, and we don't need a response because it's fine if this transaction needs recovery?

belliottsmith replied:
Basically, though in practice this is a stub for whatever we may want to do with our implementations that use this. Or, for some distributed cleanup for those nodes that don't need it - this is probably the one part of this patch that isn't "done" (but is close enough that we could probably commit as things are so we have a shared base to operate on)

aweisberg replied:
I mentioned this earlier, but for inclusive sync point I think we need to know when it is applied (because apply waits for dependencies right?) since we want the side effects of the dependencies.

belliottsmith replied:
Yep, my expectation is that the coordinator would register a listener for the application of the transaction locally. The apply at remote replicas is just a matter of cleaning up really. We do not need any remote replica to apply in order to achieve our goal, so we can return immediately upon reaching consensus and use our dependencies for our purposes.

aweisberg replied:
OK, there are two versions of this though. One is locally listening (key migration), but in the other we actually want to know it was applied at a quorum remotely (range migration)?

Or can we locally listen for the state change to globally durable?

I wonder if we will have trouble with lost notifications here. Like we may not be involved in the dependency transactions (range we don't replicate) at all so why would we receive notification of it?

belliottsmith replied:
True, for the global range migration version we might want to tweak this to support waiting for a certain level of persistence at all replicas. Is it better to add that at point of use, or would you like to add it here in this patch?

aweisberg replied:
I can do it later. I think I will also need to add the local metadata only behavior of the sync point where we already know it has been satisfied.

aweisberg commented on accord-core/src/main/java/accord/coordinate/Propose.java:
```diff
@@ -74,18 +73,19 @@ class Propose implements Callback<AcceptReply>
 
-    public static void propose(Node node, Ballot ballot, TxnId txnId, Txn txn, FullRoute<?> route,
-                               Timestamp executeAt, Deps deps, BiConsumer<Result, Throwable> callback)
+    public static void propose(Node node, Topologies topologies, Ballot ballot, TxnId txnId, Txn txn, FullRoute<?> route,
```
Unused now

aweisberg replied:
FYI still unused

aweisberg commented on accord-core/src/main/java/accord/coordinate/Propose.java:
```diff
@@ -128,12 +131,7 @@ public void onCallbackFailure(Id from, Throwable failure)
+    abstract void onAccepted();
 
     // A special version for proposing the invalidation of a transaction; only needs to succeed on one shard
     static class Invalidate implements Callback<AcceptReply>
```
Should invalidate be pulled out into a separate class now?

belliottsmith replied:
I don't think so, but happy to be told otherwise. We have a pattern of ActionX and ActionX.Invalidate - though perhaps this isn't a good pattern anyway 🤷

aweisberg commented on accord-core/src/main/java/accord/coordinate/Recover.java:
```diff
@@ -212,56 +213,63 @@ private void recover()
-                    });
+                {
+                    // must have gone through Accepted, so we must have witnessed >= Accepted for each shard
+                    Deps acceptedDeps = tryMergeAcceptedDeps();
```
We don't need to collect deps anymore? Does this mean we always enter `Recover` after reaching a quorum from each shard?

belliottsmith replied:
Yes, we require a quorum from each side for recovery. There were TODOs about using the deps we have where possible, and for this work it made sense to just do that since we need it for the exclusive sync point - which *proposes* deps, i.e. expects *precisely* those deps to be re-proposed should it fail to durably commit/apply/whathaveyou.

aweisberg commented on accord-core/src/main/java/accord/coordinate/Recover.java:
```diff
@@ -212,56 +213,63 @@ private void recover()
-                    return;
+                {
+                    Deps deps;
+                    if (txnId.rw().proposesDeps()) deps = tryMergeAcceptedDeps();
```
Why do we have this distinction here for this transaction type? (needs comment)

Seems like this is the one case where we are ok with accepted deps being null, but for a regular transaction it's an invariant that it not be null?

So a `proposeDeps` transaction must arrive here with less information, but still want to proceed to invalidation? Why is that?

belliottsmith replied:
So, this is something that is probably going to evolve. In practice we *don't* need to treat them differently today, because we do actually send deps with accept today, *and* because we are able to invalidate any transaction that hadn't reached a quorum of accepts (though we don't currently, we probably *should*, as it is consistent with our behaviour elsewhere of invalidating transactions we know were not executed by their coordinator). 

However (and this is documented, but evidently not well), standard transactions are happy to get deps that are superset of those they need, and then filter them. However, exclusive sync points want to _precisely_ exclude txnId they did not witness. We don't want extra txnId sneaking in next time around, as that wouldn't have been witnessed by the replicas that wanted to know for sure they had a complete set of txnId lower than that sync point. So for these transactions we _require_ that we re-propose precisely the same deps as the original coordinator _or_ we invalidate the transaction if they cannot be found.

belliottsmith replied:
I will of course try to express this better than I have, and in more places.

aweisberg commented on accord-core/src/main/java/accord/coordinate/Recover.java:
```diff
@@ -212,56 +213,63 @@ private void recover()
+                        return;
+                    }
+
+                    // if we propose deps, and cannot assemble a complete set, then we never reached consensus and can be invalidated
```
What does it mean to propose deps?

belliottsmith replied:
Send `Accept` with a specific set of `Deps` we want to be durably recoverable before we use them. So for these transactions we're really agreeing `Deps` rather than `executeAt` (since they always use `txnId` for that, though this is arbitrary)

belliottsmith replied:
I think this will become more confusing when we permit regular transactions to _not_ send deps, but it is already the case that we do not treat the `deps` register on replicas as durable/consistent for regular transactions. I will think about where best to document this - I will at least mention it in `Command`, and perhaps also in `RecoverOk` and `Accept`

aweisberg commented on accord-core/src/main/java/accord/local/SafeCommandStore.java:
```diff
@@ -66,7 +70,37 @@
     }
     enum TestDep { WITH, WITHOUT, ANY_DEPS }
-    enum TestKind { Ws, RorWs }
+    enum TestKind implements Predicate<Txn.Kind>
```
So with this do sync points only conflict in one direction and basically not block other transactions? Nice.

belliottsmith replied:
Yep, that's the idea

aweisberg commented on accord-core/src/main/java/accord/messages/BeginRecovery.java:
```diff
@@ -152,16 +164,18 @@ public RecoverReply reduce(RecoverReply r1, RecoverReply r2)
         if (!ok1.status.hasBeen(PreAccepted)) throw new IllegalStateException();
 
         PartialDeps deps = ok1.deps.with(ok2.deps);
+        PartialDeps acceptedDeps = ok1.deps == ok1.acceptedDeps && ok2.deps == ok2.acceptedDeps ? deps : ok1.acceptedDeps.with(ok2.acceptedDeps);
```
If one or more of them is `PartialDeps.NONE` why are we doing `with`? To make sure we pick whichever one might not be `NONE` or get double `NONE` merged?

belliottsmith replied:
If either are NONE then the operation bails out cheaply. We don't know if either of them are NONE though, they might both be full deps. Perhaps I'm misunderstanding. All I'm doing with the ternary is avoiding calling `with` twice if I can reuse the prior result.

aweisberg replied:
Ok, it took me a minute. What this is doing is creating the accepted deps only if it has to because it's not the same as `deps` when one or more has returned `NONE` for accepted deps.

aweisberg commented on accord-core/src/main/java/accord/messages/BeginRecovery.java:
```diff
@@ -216,19 +230,21 @@ public static class RecoverOk extends RecoverReply
         public final Ballot accepted;
         public final Timestamp executeAt;
         public final PartialDeps deps;
+        public final PartialDeps acceptedDeps; // only those deps that have previously been proposed
```
Good place to comment on when/why the distinction matters, but it if it can/should be covered where this is used that is also fine

belliottsmith replied:
Yep, agreed

aweisberg commented on accord-core/src/main/java/accord/messages/BeginRecovery.java:
```diff
@@ -297,7 +313,7 @@ private static Deps acceptedStartedBeforeWithoutWitnessing(SafeCommandStore comm
         try (Deps.Builder builder = Deps.builder())
         {
-            commandStore.mapReduce(keys, ranges, RorWs, STARTED_BEFORE, startedBefore, WITHOUT, startedBefore, Accepted, PreCommitted,
+            commandStore.mapReduce(keys, ranges, Any, STARTED_BEFORE, startedBefore, WITHOUT, startedBefore, Accepted, PreCommitted,
```
Why use `Any` for recovery? What is the interaction you are looking for with sync transactions?

Does recovery force them to order with sync transactions, but during regular execution we don't?

belliottsmith replied:
These collections are used for _invalidation_, not for deps or execution. We're looking here for the transactions that _invalidate_ us by _not witnessing us_. For sync points (most importantly exclusive sync points) this is the whole purpose of durably agreeing Deps - so that we know for sure we have seen all `TxnId` that can _possibly_ be agreed to execute.

aweisberg commented on accord-core/src/main/java/accord/primitives/Timestamp.java:
```diff
@@ -19,12 +19,16 @@
 
 public class Timestamp implements Comparable<Timestamp>
 {
+    private static final int REJECTED_FLAG = 0x8000;
```
Should `REJECTED_FLAG` be the same as `MERGE_FLAGS`?

belliottsmith replied:
I will comment. The idea here is that we probably don't want all flags to be mergeable, though today we only have one flag, and we do want it to propagate.

aweisberg commented on accord-core/src/main/java/accord/primitives/Txn.java:
```diff
@@ -53,6 +66,16 @@ public boolean isRead()
 
+        /**
+         * Note that it is only possible to do this for transactions whose execution time is not dependent on
+         * others, i.e. where we may safely propose executeAt = txnId regardless of when it is witnessed by
```
How does a transaction end up being that way? What transaction could safely propose a txnId of 0 at any time?

Also what is "this"? :-)

It's proposing invalidations, don't those need to properly order with what already occurred?

belliottsmith replied:
It's only possible to "propose deps". I will clean up this comment, to contain similar text to the other proposed places to explain this distinction. In this case we're essentially saying that, because this transaction doesn't actually execute, it doesn't need to agree an `executeAt`, permitting it the option of instead agreeing some `deps` from the single `PreAccept` round (since the deps after `PreAccept` are valid for `executeAt=txnId`, but not for `executeAt>txnId`)

aweisberg commented on accord-core/src/main/java/accord/primitives/Txn.java:
```diff
@@ -32,13 +32,26 @@
+         * A pseudo-transaction whose deps represent the complete set of transactions that may execute before it,
+         * without interfering with their execution.
+         *
+         * A SyncPoint is unique in that it agrees not only an executeAt but also a precise collection of dependencies,
```
Precise collection needs elaboration, what you mean is that after `Preaccept` we have a precise set because we are fine with extra dependencies?

Don't all transactions determine a precise set of dependencies in order to determine their `executeAt`?

belliottsmith replied:
Nope, most transactions do not _agree_ their deps, and indeed for those that agree with 1RT consensus there are no durable deps to recover (and we intend to stop supplying them for accept too). But for SyncPoint, for the reasons stated elsewhere, this is fine because we always have `executeAt=txnId` and this is fine because we don't "execute" them.

aweisberg replied:
I still don't follow something basic. What does agree mean?

Maybe we need to step up a level and explain how a sync point achieves its objective where when a result is returned all the dependencies of this txn have reached applied.

Maybe I am crazy, but it looks like the callback fires after `Propose` completes, and doesn't wait for the dependencies to actually apply?

I am going to reread again today and see if I can get to a better understanding.

aweisberg replied:
The reason the dependencies need to be applied is that the use case is needing the side effects of the dependencies to already have occurred for migration.

belliottsmith replied:
Agree means we have a quorum of accepted proposals. Ie we have reached a durable decision. I think this is quite normal terminology? At least that’s how I talk about consensus

aweisberg replied:
I am figuring through how regular transactions don't agree on their deps, so with 1RT yeah that makes sense.

There are no changes to how `Accept` handles deps so I was trying to understand what it means for these transactions to agree on deps, but regular transactions don't. So one different is we always run accept, and I guess in the future you mean that accept won't persist dependencies for regular transactions (recovery will go find them again?), but we will continue to persist them for barriers so when they recover the precise set of dependencies is returned.

belliottsmith replied:
Yes, but it’s slightly more than that: since an Accept round for a normal transaction means it has (likely) taken a later executeAt the dependencies aren’t enough for execution as they don’t include the full set of deps for the higher executeAt - for that we use the dependencies returned in the Acceot response

aweisberg commented on accord-core/src/main/java/accord/primitives/Txn.java:
```diff
@@ -32,13 +32,26 @@
+         * without interfering with their execution.
+         *
+         * A SyncPoint is unique in that it agrees not only an executeAt but also a precise collection of dependencies,
+         * so that is effect on invalidation of earlier transactions is durable. This is most useful for ExclusiveSyncPoint.
```
This last half really confuses me, `SyncPoint` shouldn't invalidate?

belliottsmith replied:
An `ExclusiveSyncPoint` invalidates all `TxnId` lower than it that aren't witnessed by it.

aweisberg replied:
Ah got it, mixed up that `SyncPointer` includes both.

aweisberg reviewed (changes requested):
Bunch of questions and nits

Still reviewing `ReducingIntervalMap` and `ReducingRangeMap`

aweisberg commented on accord-core/src/main/java/accord/utils/ReducingRangeMap.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    /**
+     * returns a copy of this LinearRangeMap limited to the ranges supplied, with all other ranges reporting Ballot.none()
```
It's not a ballot anymore

belliottsmith commented:
Ok, have addressed the last nits. Merge on hold until immutable state lands.

aweisberg committed:
ExclusiveSync transactions

Port over PaxosRepairHistory

ExclusiveSync transactions

testing

make sure sync points impact recovery of other transactions

feedback

belliottsmith committed:
durable+transient listeners

belliottsmith committed:
backport from bootstrap various changes

belliottsmith commented:
I've rebased and back ported some relevant changes from the bootstrap PR (plus some changes not super tied to either). I've also brought in some changes to listener handling. I plan to merge this to cep-15-accord alongside bootstrap, to simplify things, but would be good to get this review signed off once you're ready.
